contrib/colab/DC-GAN.ipynb

```diff
@@ -0,0 +1,554 @@
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "EZ45r5UvQhxe",
+        "colab": {}
+      },
+      "source": [
+        "def real_data_target(size, device):\n",
+        "    '''\n",
+        "    Tensor containing ones, with shape = size\n",
+        "    '''\n",
+        "    data = torch.ones(size, 1)\n",
+        "    return data.to(device)\n",
+        "\n",
+        "def fake_data_target(size, device):\n",
+        "    '''\n",
+        "    Tensor containing zeros, with shape = size\n",
+        "    '''\n",
+        "    data = torch.zeros(size, 1)\n",
+        "    return data.to(device)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "2ofS8wztQhxi",
+        "colab": {}
+      },
+      "source": [
+        "# Noise\n",
+        "def noise(size, device):\n",
+        "    n = torch.randn(size, 100)\n",
+        "    return n.to(device)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "colab_type": "text",
+        "id": "ynBtejcpQhxj"
+      },
+      "source": [
+        "# Note on the use of .detach() method\n",
+        "\n",
+        "You will notice in the following code snippet that when the generator is used to create the `fake_data`, .detach() for the discriminator training step, the .detach call is used to create a new view of the fake_data tensor for which the operations will not be recorded for gradient computation.<br>\n",
+        "\n",
+        "Since fake_date is an output of an nn.module, by default, pytorch will record all the operations performed on this tensor during the forward pass as DAG. And after the backward pass these DAG and corresponding operations are cleared (unless `retain_graph=True`).\n",
+        "Therefore such a tensor can be part of only one cone of logic where the forward and backward pass is done. If there are two loss function where this tensor is used and backward pass is performed on these two function (or even sum of the functions) for the second backward pass the operations DAG will not be found, leading to an error.\n",
+        "\n",
+        "The second place, where detach() is used is when a numpy() call is to be made to tensor (for plotting purposes). Pytorch also requires that requires_grad should not be true on these tensor. (Ref: `\n",
+        "RuntimeError: Can't call numpy() on Variable that requires grad. Use var.detach().numpy() instead.`)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "colab_type": "code",
+        "id": "OFzyeN_GQhxk",
+        "colab": {}
+      },
+      "source": [
+        "def train_gan(rank):\n",
+        "    torch.manual_seed(1)\n",
+        "    SERIAL_EXEC = xmp.MpSerialExecutor()\n",
```

Comment:
This needs to be created at global scope.

Reply:
Should I place it just outside train_gan function? else it would be too far removed from where it  is going to be used. else I can place it where FLAGS are defined? suggestions?

Reply:
Got it nvm.

Reply:
Done.